xds/src/main/java/io/grpc/xds/RingHashLoadBalancer.java

```diff
@@ -0,0 +1,455 @@
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    if (currentState != READY) {
+      helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    for (Subchannel subchannel : subchannels.values()) {
+      shutdownSubchannel(subchannel);
+    }
+  }
+
+  private void updateBalancingState() {
+    checkState(!subchannels.isEmpty(), "no subchannel has been created");
+    ConnectivityState overallState = aggregateState(subchannels.values());
+    RingHashPicker picker = new RingHashPicker(syncContext, ring, subchannels);
+    // TODO(chengyuanzhang): avoid unnecessary reprocess caused by duplicated server addr updates
+    helper.updateBalancingState(overallState, picker);
+    currentState = overallState;
+  }
+
+  private void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
+    if (subchannels.get(stripAttrs(subchannel.getAddresses())) != subchannel) {
+      return;
+    }
+    Ref<ConnectivityStateInfo> subchannelStateRef = getSubchannelStateInfoRef(subchannel);
+
+    // Don't proactively reconnect if the subchannel enters IDLE, even if previously was connected.
+    // If the subchannel was previously in TRANSIENT_FAILURE, it is considered to stay in
+    // TRANSIENT_FAILURE until it becomes READY.
+    if (subchannelStateRef.value.getState() == TRANSIENT_FAILURE) {
+      if (stateInfo.getState() == CONNECTING || stateInfo.getState() == IDLE) {
+        return;
+      }
+    }
+    subchannelStateRef.value = stateInfo;
+    updateBalancingState();
+  }
+
+  /**
+   * Aggregates the connectivity states of a group of subchannels for overall connectivity state.
+   *
+   * <p>Aggregation rules (in order of dominance):
+   * <ol>
+   *   <li>If there is at least one subchannel in READY state, overall state is READY</li>
+   *   <li>If there are <em>more than 2</em> subchannels in TRANSIENT_FAILURE, overall state is
+   *   TRANSIENT_FAILURE</li>
+   *   <li>If there is at least one subchannel in CONNECTING state, overall state is
+   *   CONNECTING</li>
+   *   <li>If there is at least one subchannel in IDLE state, overall state is IDLE</li>
+   *   <li>Otherwise, overall state is TRANSIENT_FAILURE</li>
+   * </ol>
+   */
+  private static ConnectivityState aggregateState(Iterable<Subchannel> subchannels) {
+    int failureCount = 0;
+    boolean hasIdle = false;
+    boolean hasConnecting = false;
+    for (Subchannel subchannel : subchannels) {
+      ConnectivityState state = getSubchannelStateInfoRef(subchannel).value.getState();
+      if (state == READY) {
+        return state;
+      }
+      if (state == TRANSIENT_FAILURE) {
+        failureCount++;
+      } else if (state == CONNECTING) {
+        hasConnecting = true;
+      } else if (state == IDLE) {
+        hasIdle = true;
+      }
+    }
+    if (failureCount > 2) {
```

Comment:
TODO: this should be `>=2`. That is, the overall state should be `TRANSIENT_FAILURE` if there are exactly two subchannels in TRANSIENT_FAILURE. When this ring hash LB policy is under the priority LB policy, the logic for kicking off connection for the next first IDLE subchanel when the _first two_ subchannels are in TF will not be used. 

However, ring hash LB policy will be a general purpose LB policy, it can be used directly by the Channel. In that case, Channel will still be able to reprocess even if the overall state is TF. We still want the priority policy to fail over quickly, so do not change this criteria from 2 to 3.

See discussions on https://docs.google.com/document/d/10J27vUqOYH8du6XZLcnmwOOSwO9r3HjIe_Jybn6AbVE/edit?disco=AAAAIIqf0Tk

Reply:
Updated.